go/core/background_action.go

```diff
@@ -0,0 +1,188 @@
+	Supports       any                // Capabilities this model supports
+	ConfigSchema   *jsonschema.Schema // Custom options schema for this model
+	Label          string             // Descriptive name for this model
+	SupportsCancel bool               // Whether the model supports cancellation
+}
+
+// BackgroundStartFunc starts a background operation
+type BackgroundStartFunc[In, Out any] func(ctx context.Context, input In) (*Operation, error)
+
+// BackgroundCheckFunc checks the status of a background operation
+type BackgroundCheckFunc[Out any] func(ctx context.Context, operation *Operation) (*Operation, error)
+
+// BackgroundCancelFunc cancels a background operation
+type BackgroundCancelFunc[Out any] func(ctx context.Context, operation *Operation) (*Operation, error)
+
+// BackgroundAction interface represents a background action
+type BackgroundAction[In, Out any] interface {
+	Start(ctx context.Context, input In) (*Operation, error)
+	Check(ctx context.Context, operation *Operation) (*Operation, error)
+	Cancel(ctx context.Context, operation *Operation) (*Operation, error)
+	SupportsCancel() bool
+	Name() string
+}
+
+// backgroundActionImpl implements BackgroundAction
+type backgroundActionImpl[In, Out any] struct {
+	startAction    *ActionDef[In, *Operation, struct{}]
+	checkAction    *ActionDef[*Operation, *Operation, struct{}]
+	cancelAction   *ActionDef[*Operation, *Operation, struct{}]
+	name           string
+	supportsCancel bool
+}
+
+// Start initiates a background operation
+func (b *backgroundActionImpl[In, Out]) Start(ctx context.Context, input In) (*Operation, error) {
+	return b.startAction.Run(ctx, input, nil)
+}
+
+// Check polls the status of a background operation
+func (b *backgroundActionImpl[In, Out]) Check(ctx context.Context, operation *Operation) (*Operation, error) {
+	return b.checkAction.Run(ctx, operation, nil)
+}
+
+// Cancel attempts to cancel a background operation
+func (b *backgroundActionImpl[In, Out]) Cancel(ctx context.Context, operation *Operation) (*Operation, error) {
+	if !b.supportsCancel || b.cancelAction == nil {
+		return nil, NewError(UNIMPLEMENTED, "cancel operation not supported")
+	}
+	return b.cancelAction.Run(ctx, operation, nil)
+}
+
+// SupportsCancel returns whether the action supports cancellation
+func (b *backgroundActionImpl[In, Out]) SupportsCancel() bool {
+	return b.supportsCancel
+}
+
+// Name returns the action name
+func (b *backgroundActionImpl[In, Out]) Name() string {
+	return b.name
+}
+
+// DefineBackgroundAction creates and registers a background action with three component actions
+func DefineBackgroundAction[In, Out any](
+	r *registry.Registry,
+	provider, name string,
+	config BackgroundModelOptions,
+	metadata map[string]any,
+	start BackgroundStartFunc[In, Out],
+	check BackgroundCheckFunc[Out], // Function to check operation status
+	cancel BackgroundCancelFunc[Out], // Optional function to cancel operation
```

Comment:
Move all of these to `BackgroundModelOptions`. The arg list should be `provider`, `name`, `opts`.